File: talleres/taller07/LinkedList.java
/**
 * Taller 7
 * 
 * LinkedList implementation in Java.
 * 
 * @author  Juan Pablo Vidal
 *          Alejandro Murillo
 */
public class LinkedList {
    
    /**
     * Execution of some test cases.
     * @param args 
     */
      public static void main(String args[]) {
          
          LinkedList list = new LinkedList();
          list.add(1);
          list.add(2);
          list.add(3);
          list.add(4);
          
          list.print();
          
          list.add(10);
          
          list.print();
          
          list.addAtStart(3);
          
          list.print();
          list.printInverse();
          try {
                list.print(3);
          } catch (Exception error) {
                System.out.println(error);
          }
          
          list.deleteEnd();
          list.print();
          
          list.deleteStart();
          list.print();
          
          try {
            System.out.println(maximum(list));
          } catch (Exception error) {
              System.out.println(error);
          }
          
          LinkedList list2 = new LinkedList();
          list2.add(3);
          list2.add(9);
          
          System.out.println(list.compareTo(list2));
      }
      
      /**
       * The first node of the LinkedList.
       */
      Node start;
      /**
       * LinkedList's size.
       */
      int size;
       
      /**
       * List Nodes.
       */
       private class Node {    
           int data;     
           Node next; 
           
           public Node(int data){    
               this.data = data;
           }
       }  
       
       /**
        * Creates a new LinkedList
        * and initializes the first
        * node to null.
        */
       public LinkedList(){ 
           start= null;
       } 
       
       /**
        * Adds an element at the start of 
        * the LinkedList.
        * @param n 
        */
       public void addAtStart(int n) {  
           Node newNode = new Node(n);
           
           if (start != null) {
               newNode.next = start; 
               start= newNode;      
           } 
           else { 
               start = newNode;
           }
           
           this.size += 1;
       }
       
       /**
        * Prints the elements of the list
        * on console.
        */
       public void print() {
           
           String elements = "[";
           
           Node n = this.start;
           while (n != null) { 
               elements += n.data +  ", ";
               n = n.next;
           }
           System.out.println(elements.substring(0, elements.length() - 2) + "]");
        } 
       
       /**
        * Adds an element to the end of the list.
        * 
        * @param n 
        */
       public void add(int n) {
           
           if (this.start == null) {
               this.start = new Node(n);
               return;
           }
                       
           Node node = this.start;
           while (node.next != null) {
               node = node.next;
           }
           node.next = new Node(n);
           this.size += 1;
       }
       
       /**
        * Prints the elements of the
        * list in reverse order.
        */
       public void printInverse() { 
           
           String elements = "";
           Node n = this.start;
           while (n != null) { 
               elements = ", " + n.data + elements;
               n = n.next;
           }
           System.out.println("[" + elements.substring(2) + "]");
           
       }
       
       /**
        * Prints the element at position n.
        * 
        * @param n
        * @throws Exception 
        */
       public void print(int n) throws Exception { 
           
           if (n >= size) {
               throw new Exception("Array Out of Bound Exception");
           }
           
           Node node = this.start;
           for (int i = 0; i < n && node != null; ++i) {
               node = node.next;
           }
           System.out.println(node.data);
       }  
       
       /**
        * Deletes the last element on
        * the LinkedList.
        */
       public void deleteEnd() { 
           
           Node node = this.start;
           while (node.next.next != null) {
               node = node.next;
           }
           node.next = null;
           this.size -= 1;
       }
       
       /**
        * Deletes the first element of the 
        * LinkedList.
        */
       public void deleteStart() { 
           
           this.start = this.start.next;
           this.size -= 1;
       }
       
       /**
        * Recursively search for the maximum
        * element.
        * 
        * @param node
        * @return 
        */
       private static int maximoAux(Node node) {
           if (node.next == null) return node.data;
           
           if (node.data > node.next.data)
               node.next.data = node.data;
           
           return maximoAux(node.next);
       }
       
       /**
        * Returns the maximum element of a
        * LinkedList.
        * 
        * @param lista
        * @return
        * @throws Exception 
        */
       public static int maximum(LinkedList lista) throws Exception {
           if (lista.start == null)
               throw new Exception("Null LinkedList: Behaviour undefined.");
           
           return maximoAux(lista.start);
       }
       
       /**
        * Compares this LinkedList with 
        * the one sent as a parameter.
        * If they are equal returns true,
        * false otherwise.
        * 
        * @param list
        * @return 
        */
       public boolean compareTo(LinkedList list) {  
           
           Node nodeThis = this.start;
           Node nodeThat = list.start;
           while (nodeThis != null) { 
               if (nodeThis.data != nodeThat.data || (nodeThis.next == null && nodeThat.next != null) || (nodeThat.next == null && nodeThis.next != null)) 
                   return false;
               nodeThis = nodeThis.next;
               nodeThat = nodeThat.next;        
           }
           
           return true;
       }  
}

File: proyecto/codigo/LinkedList.py
class LinkedList:
    """
    Implementation of a simple
    LinkedList.
    @Author: Alejandro Murillo
    """

    class Node:
        """
        Implementation of the
        LinkedList's node.
        """

        def __init__(self, data):
            """
            Creates a new Node.
            :param data: The node's data.
            """
            self.data = data
            self.next = None
            self.previous = None

        def get_data(self):
            return self.data

        def __str__(self):
            return str(self.data)

    def __init__(self):
        """
        Creates a new LinkedList.
        Initializes the Head and Tail node
        to None; and its size to 0.
        """
        self.head = None
        self.tail = None
        self.size = 0

    def add_at_start(self, data):
        """
        Adds an element at the start of the
        LinkedList.
        :param data: The new Node data.
        :return: Void
        """

        new_node = self.Node(data)

        if self.head is not None:
            self.head.previous = new_node
            new_node.next = self.head

        self.head = new_node
        self.size += 1

        if self.size == 1:
            self.tail = self.head

    def add(self, data):
        """
        Adds an element to the end of
        the list.
        :param data: The new Node data.
        :return: Void
        """

        if self.head is None:
            self.add_at_start(data)
        else:
            new_node = self.Node(data)
            new_node.previous = self.tail
            self.tail.next = new_node
            self.tail = new_node
            self.size += 1

    def delete_end(self):
        """
        Deletes the last element on the
        LinkedList.
        :return: Void
        """
        self.tail = self.tail.previous
        self.tail.next = None
        self.size -= 1

    def delete_start(self):
        """
        Deletes the first element
        on the LinkedList.
        :return: Void
        """
        self.head = self.head.next
        self.head.previous = None
        self.size -= 1

    def get(self, i):
        """
        Returns the ith linked element.
        :param i: Element's index.
        :return: Node
        """

        if i >= self.size or i < 0:
            return IndexError("Unavailable element at the specified index.")

        if self.head is not None:
            return None

        iteration = 0
        current = self.head

        while iteration != i and current.next is not None:
            current = current.next
            iteration += 1

        return current

    def print_reversed(self):
        """
        Prints the list in reverse order.
        :return: Void
        """
        current = self.tail
        s = str(current.data)
        while current.previous is not None:
            current = current.previous
            s += ", " + str(current.data)

        print(s)

    def __str__(self):
        if self.head is None:
            return "[]"
        current = self.head
        s = "[" + str(current.data)
        while current.next is not None:
            current = current.next
            s += ", " + str(current.data)

        return s + "]"

    def __contains__(self, item):

        current = self.head
        while current is not None:
            if current.data == item:
                return True
            current = current.next

        return False

    def __len__(self):
        return self.size
